Issue webhook payload

thehowl opened the pull request:
This PR prepares for an upcoming pull request on the main repository that aims to implement https://github.com/go-gitea/gitea/issues/132, adding webhook for issue events.

A small API breakage would be introduced in the SDK (though the API JSON objects' structure would remain untouched, if not for some shifting of the position of the fields). As most fields between the PullRequestPayload and the IssuePayload would be the same, the idea is to move the fields to a common BaseIssuePayload, although this would bring a small breakage to when constructing the struct of a PullRequestPayload:

```go
// No longer possible after this PR
pr := &gitea.PullRequestPayload{
	Index: 64, 
}
```

However, accessing the struct fields would still be possible after the struct is initialised.

```go
pr := &gitea.PullRequestPayload{}
pr.Index = 64
_ = pr.Index
```

This would be, however, a very small issue, as the PullRequestPayload is only initialised by Gitea itself ([see this github search](https://github.com/search?utf8=%E2%9C%93&q=code.gitea.io%2Fsdk+PullRequestPayload&type=Code&ref=searchresults)), but Gitea vendors its dependencies anyway, so this would break only in the event of a vendor update with no update to models/issue.go's struct initialisations.

thehowl committed:
Add Issue webhook payload

thehowl committed:
Add copyright header

thehowl committed:
Whoops, wrote license header in wrong file...

thehowl committed:
Move fields of PR/Issue payload to BaseIssuePayload

lunny commented on gitea/hook.go:
```diff
@@ -2,6 +2,10 @@
 // Use of this source code is governed by a MIT-style
 // license that can be found in the LICENSE file.
 
+// Copyright 2017 The Gitea Authors. All rights reserved.
```
This line is the only need and put it after the first line.

thehowl committed:
Fix license header

lunny commented:
How about use the duplicated struct between `Issue` and `Pull` for the compitable?

thehowl commented:
For the computable? I'm sorry, I did not understand. Could you elaborate?

lunny commented on gitea/hook.go:
```diff
@@ -277,17 +295,39 @@ const (
+	Issue *Issue `json:"issue"`
+}
+
+// JSONPayload FIXME
```
comment

lunny commented:
I mean maybe remove `BaseIssuePayload` and copy the fields to `IssuePayload` and `PullPayload` is better.

thehowl commented:
Yeah, that was how I originally thought about doing it (and the safest way, since it doesn't cause any breakage). However, it's a matter of convenience and code de-duplication.

I'm gonna do as you said anyway, mostly because of what @tboerger said on Gitter.

thehowl committed:
remove BaseIssuePayload

thehowl committed:
Add comment to IssuePayload.JSONPayload

thehowl committed:
Fix comment mentioning IssuePayload

thehowl committed:
Reorder structs

lunny commented:
LGTM

thehowl committed:
Mimic GitHub webhooks

Since the Gitea webhooks currently mimic for a big part the way GitHub does webhook, issue comments HookIssueActions have been removed to be added again at a later stage and milestone_set and milestone_cleared have been changed to milestoned and demilestoned, respectively.

thehowl commented:
I've changed the HookIssueAction names of the ones I added, and removed the comment ones, to better mimic the GitHub webhooks. Issue comments will come at a later stage (and PR).

thehowl committed:
Change also constant names

tboerger commented:
LGTM
